File: backend/app/vision/landmark_processor.py
def extract_features(landmarks):
    """
    Extracts normalized features from 21 MediaPipe hand landmarks.
    Each landmark should be a dict or object with x, y, z attributes.
    """
    if len(landmarks) != 21:
        raise ValueError("Expected exactly 21 landmarks.")
        
    # Handle both dict and object access
    def get_coord(lm, coord):
        if isinstance(lm, dict):
            return lm.get(coord, 0.0)
        return getattr(lm, coord, 0.0)

    # Translate relative to wrist (landmark 0)
    wrist_x = get_coord(landmarks[0], 'x')
    wrist_y = get_coord(landmarks[0], 'y')
    wrist_z = get_coord(landmarks[0], 'z')
    
    translated = []
    max_val = 0.0
    
    for lm in landmarks:
        dx = get_coord(lm, 'x') - wrist_x
        dy = get_coord(lm, 'y') - wrist_y
        dz = get_coord(lm, 'z') - wrist_z
        
        translated.append((dx, dy, dz))
        max_val = max(max_val, abs(dx), abs(dy), abs(dz))
        
    if max_val == 0.0:
        max_val = 1.0
        
    features = []
    for dx, dy, dz in translated:
        features.extend([dx / max_val, dy / max_val, dz / max_val])
        
    return features

File: frontend/src/utils/landmarkProcessor.js
/**
 * Extracts normalized features from 21 MediaPipe hand landmarks.
 * Must produce the exact same output as the Python backend logic.
 * 
 * @param {Array} landmarks Array of 21 objects with {x, y, z}
 * @returns {Array} 1D array of 63 normalized float features
 */
export function extractFeatures(landmarks) {
  if (!landmarks || landmarks.length !== 21) {
    throw new Error("Expected exactly 21 landmarks.");
  }

  // Translate relative to wrist (landmark 0)
  const wristX = landmarks[0].x || 0.0;
  const wristY = landmarks[0].y || 0.0;
  const wristZ = landmarks[0].z || 0.0;

  const translated = [];
  let maxVal = 0.0;

  for (let i = 0; i < landmarks.length; i++) {
    const lm = landmarks[i];
    const dx = (lm.x || 0.0) - wristX;
    const dy = (lm.y || 0.0) - wristY;
    const dz = (lm.z || 0.0) - wristZ;

    translated.push([dx, dy, dz]);
    maxVal = Math.max(maxVal, Math.abs(dx), Math.abs(dy), Math.abs(dz));
  }

  if (maxVal === 0.0) {
    maxVal = 1.0;
  }

  const features = [];
  for (let i = 0; i < translated.length; i++) {
    const [dx, dy, dz] = translated[i];
    features.push(dx / maxVal, dy / maxVal, dz / maxVal);
  }

  return features;
}
